src/usr/share/netcheck/networkmanagerhelper.py

```diff
@@ -0,0 +1,202 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""NetworkManagerHelper provides a readable layer of abstraction for the
+python-networkmanager class.
+"""
+
+__all__ = ['NetworkManagerHelper']
+__author__ = 'Andrew Klapp and Joel Allen Luellwitz'
+__version__ = '0.8'
+
+import datetime
+import logging
+import time
+import NetworkManager
+
+NETWORKMANAGER_ACTIVATION_CHECK_DELAY = 0.1
+
+NM_CONNECTION_ACTIVATING = 0
+NM_CONNECTION_ACTIVE = 1
+NM_CONNECTION_DISCONNECTED = 2
+
+class DeviceNotFoundException(Exception):
+    """Raised when an interface name passed to NetworkManagerHelper is not found."""
+
+class NetworkManagerHelper:
+    """NetworkManagerHelper abstracts away some of the messy details of the NetworkManager
+    Dbus API.
+    """
+
+    def __init__(self, config):
+
+        self.logger = logging.getLogger(__name__)
+        self.network_activation_timeout = config['network_activation_timeout']
+        self.wired_network_name = config['wired_network_name']
+
+        self.network_id_table = self._build_network_id_table()
+        self._get_device_objects(config['wired_interface_name'],
+            config['wifi_interface_name'])
+
+    def activate_network(self, network_id):
+        """Tells NetworkManager to activate a network with the supplied network_id.
+        Returns True if there are no errors, False otherwise.
+        """
+        success = False
+        connection = self.network_id_table[network_id]
+        network_device = self._get_device_for_connection(connection)
+
+        networkmanager_output = NetworkManager.NetworkManager.ActivateConnection(connection, network_device, '/')
+        self._run_proxy_call(networkmanager_output)
+        success = self._wait_for_connection(connection)
+
+        return success
+
+    def get_network_ip(self, network_id):
+        """Attempts to retrieve the ip address associated with the given network id. If it
+        is unable to, it returns None."""
+        # I decided not to throw an exception here because the intended caller would end up
+        #   simply using it for flow control.
+        ip_address = None
+
+        connection = self.network_id_table[network_id]
+
+        if self._wait_for_connection(connection):
+            device = self._get_device_for_connection(connection)
+            ip_address = device.Ip4Config.AddressData[0]['address']
+        else:
+            self.logger.warning('Attempted to get IP address for network that is not connected.')
```

Comment:
I'd probably assign None to ip_address down here.

Reply:
I keep forgetting this isn't Java with the 'final' keyword. It might be better to leave it where you have it.